File: mac/Sources/BridgeCore/Pairing.swift
import Foundation
import CryptoKit
import DeviceLinkProtocol

public struct DeviceIdentity: Codable, Equatable {
    public let deviceId: String
    public let deviceName: String
    public let publicKeyB64: String
}

public struct PairedDevice: Codable, Equatable {
    public let deviceId: String
    public let deviceName: String
    public let publicKeyB64: String
    public let fingerprint: String
    public var host: String = ""
    public var port: Int = 0
}

public struct QrPayload: Codable, Equatable {
    public let deviceId: String
    public let deviceName: String
    public let publicKeyB64: String
    public let fingerprint: String
    public let host: String
    public let port: Int
}

/// Pairing & trust (U2). EC P-256 via CryptoKit; trust-on-first-use cert pinning recorded in SecureStore.
public final class PairingManager {
    private let store: SecureStore
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    public init(store: SecureStore) { self.store = store }

    public func fingerprint(of publicKeyB64: String) -> String {
        guard let data = Data(base64Encoded: publicKeyB64) else { return "" }
        return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined(separator: ":")
    }

    public func generateIdentity(_ deviceName: String) -> DeviceIdentity {
        let priv = P256.Signing.PrivateKey()
        let pubB64 = priv.publicKey.rawRepresentation.base64EncodedString()
        let identity = DeviceIdentity(deviceId: UUID().uuidString, deviceName: deviceName, publicKeyB64: pubB64)
        store.put(Keys.identity, (try? String(data: encoder.encode(identity), encoding: .utf8)) ?? "")
        store.put(Keys.privateKey, priv.rawRepresentation.base64EncodedString())
        return identity
    }

    public func loadIdentity() -> DeviceIdentity? {
        guard let s = store.get(Keys.identity), let d = s.data(using: .utf8) else { return nil }
        return try? decoder.decode(DeviceIdentity.self, from: d)
    }

    public func createPairingQr(_ identity: DeviceIdentity, host: String, port: Int) -> String {
        let payload = QrPayload(
            deviceId: identity.deviceId, deviceName: identity.deviceName,
            publicKeyB64: identity.publicKeyB64, fingerprint: fingerprint(of: identity.publicKeyB64),
            host: host, port: port
        )
        return (try? String(data: encoder.encode(payload), encoding: .utf8)) ?? ""
    }

    public enum PairingError: Error, Equatable { case malformed, fingerprintMismatch }

    @discardableResult
    public func consumePairingQr(_ qr: String) throws -> PairedDevice {
        guard let data = qr.data(using: .utf8), let payload = try? decoder.decode(QrPayload.self, from: data) else {
            throw PairingError.malformed
        }
        if fingerprint(of: payload.publicKeyB64) != payload.fingerprint {
            LinkLogger.securityEvent("pair_fingerprint_mismatch", ["deviceId": payload.deviceId])
            throw PairingError.fingerprintMismatch
        }
        let peer = PairedDevice(
            deviceId: payload.deviceId, deviceName: payload.deviceName, publicKeyB64: payload.publicKeyB64,
            fingerprint: payload.fingerprint, host: payload.host, port: payload.port
        )
        pin(peer)
        return peer
    }

    public func pin(_ peer: PairedDevice) {
        let list = listPaired().filter { $0.deviceId != peer.deviceId } + [peer]
        savePaired(list)
        LinkLogger.info("peer_pinned", ["deviceId": peer.deviceId])
    }

    public func listPaired() -> [PairedDevice] {
        guard let s = store.get(Keys.paired), let d = s.data(using: .utf8) else { return [] }
        return (try? decoder.decode([PairedDevice].self, from: d)) ?? []
    }

    public func unpair(_ deviceId: String) {
        savePaired(listPaired().filter { $0.deviceId != deviceId })
        LinkLogger.info("peer_unpaired", ["deviceId": deviceId])
    }

    public func isPinned(_ fingerprint: String) -> Bool { listPaired().contains { $0.fingerprint == fingerprint } }

    private func savePaired(_ list: [PairedDevice]) {
        store.put(Keys.paired, (try? String(data: encoder.encode(list), encoding: .utf8)) ?? "[]")
    }

    private enum Keys {
        static let identity = "identity"
        static let privateKey = "identity.private"
        static let paired = "paired.devices"
    }
}

File: android/app/src/main/kotlin/com/androidbridge/core/Pairing.kt
package com.androidbridge.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.security.KeyPairGenerator
import java.security.MessageDigest
import java.util.Base64
import java.util.UUID

private val pairingJson = Json { ignoreUnknownKeys = true; encodeDefaults = true }

/** This device's stable identity. The private key is held in [SecureStore]; only public material travels. */
@Serializable
data class DeviceIdentity(
    val deviceId: String,
    val deviceName: String,
    val publicKeyB64: String,
)

/** A peer this device trusts (pinned at pairing time — trust-on-first-use). */
@Serializable
data class PairedDevice(
    val deviceId: String,
    val deviceName: String,
    val publicKeyB64: String,
    val fingerprint: String,
    val host: String = "",
    val port: Int = 0,
)

/** The payload encoded into the pairing QR code. */
@Serializable
data class QrPayload(
    val deviceId: String,
    val deviceName: String,
    val publicKeyB64: String,
    val fingerprint: String,
    val host: String,
    val port: Int,
)

/**
 * Pairing & trust (U2). Generates a per-device keypair, builds/consumes pairing QR payloads, and
 * pins peers via [SecureStore]. Crypto uses the JDK (EC P-256) so the logic is JVM-testable;
 * full X.509/mTLS wiring is layered on top by the connection unit.
 */
class PairingManager(private val store: SecureStore) {

    fun fingerprintOf(publicKeyB64: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(Base64.getDecoder().decode(publicKeyB64))
        return digest.joinToString(":") { "%02x".format(it) }
    }

    fun generateIdentity(deviceName: String): DeviceIdentity {
        val kpg = KeyPairGenerator.getInstance("EC").apply { initialize(256) }
        val kp = kpg.generateKeyPair()
        val pubB64 = Base64.getEncoder().encodeToString(kp.public.encoded)
        val privB64 = Base64.getEncoder().encodeToString(kp.private.encoded)
        val identity = DeviceIdentity(deviceId = UUID.randomUUID().toString(), deviceName = deviceName, publicKeyB64 = pubB64)
        store.put(KEY_IDENTITY, pairingJson.encodeToString(DeviceIdentity.serializer(), identity))
        store.put(KEY_PRIVATE, privB64)
        return identity
    }

    fun loadIdentity(): DeviceIdentity? =
        store.get(KEY_IDENTITY)?.let { pairingJson.decodeFromString(DeviceIdentity.serializer(), it) }

    fun createPairingQr(identity: DeviceIdentity, host: String, port: Int): String {
        val payload = QrPayload(
            deviceId = identity.deviceId,
            deviceName = identity.deviceName,
            publicKeyB64 = identity.publicKeyB64,
            fingerprint = fingerprintOf(identity.publicKeyB64),
            host = host,
            port = port,
        )
        return pairingJson.encodeToString(QrPayload.serializer(), payload)
    }

    /** Parse + pin a scanned QR payload (trust-on-first-use). Returns the pinned peer. */
    fun consumePairingQr(qr: String): PairedDevice {
        val payload = pairingJson.decodeFromString(QrPayload.serializer(), qr)
        val expected = fingerprintOf(payload.publicKeyB64)
        if (expected != payload.fingerprint) {
            LinkLogger.securityEvent("pair_fingerprint_mismatch", mapOf("deviceId" to payload.deviceId))
            throw IllegalArgumentException("fingerprint mismatch")
        }
        val peer = PairedDevice(
            deviceId = payload.deviceId,
            deviceName = payload.deviceName,
            publicKeyB64 = payload.publicKeyB64,
            fingerprint = payload.fingerprint,
            host = payload.host,
            port = payload.port,
        )
        pinPeer(peer)
        return peer
    }

    fun pinPeer(peer: PairedDevice) {
        val list = listPaired().filter { it.deviceId != peer.deviceId } + peer
        store.put(KEY_PAIRED, pairingJson.encodeToString(PairedListSurrogate.serializer(), PairedListSurrogate(list)))
        LinkLogger.info("peer_pinned", mapOf("deviceId" to peer.deviceId))
    }

    fun listPaired(): List<PairedDevice> =
        store.get(KEY_PAIRED)?.let {
            pairingJson.decodeFromString(PairedListSurrogate.serializer(), it).devices
        } ?: emptyList()

    fun unpair(deviceId: String) {
        val list = listPaired().filter { it.deviceId != deviceId }
        store.put(KEY_PAIRED, pairingJson.encodeToString(PairedListSurrogate.serializer(), PairedListSurrogate(list)))
        LinkLogger.info("peer_unpaired", mapOf("deviceId" to deviceId))
    }

    fun isPinned(fingerprint: String): Boolean = listPaired().any { it.fingerprint == fingerprint }

    @Serializable
    private data class PairedListSurrogate(val devices: List<PairedDevice>)

    companion object {
        private const val KEY_IDENTITY = "identity"
        private const val KEY_PRIVATE = "identity.private"
        private const val KEY_PAIRED = "paired.devices"
    }
}
